src/geoarrow.cpp

```diff
@@ -71,7 +71,178 @@ py::array_t<PyObjectGeography> from_geoarrow(py::object input,
+            }
+            free(array);
+        };
+        py::capsule array_capsule{c_array, "arrow_array", array_cleanup};
+
+        constexpr auto schema_cleanup = [](void* ptr) noexcept {
+            auto schema = static_cast<ArrowSchema*>(ptr);
+            if (schema->release != nullptr) {
+                schema->release(schema);
+            }
+            free(schema);
+        };
+        py::capsule schema_capsule{c_schema, "arrow_schema", schema_cleanup};
+
+        return py::make_tuple(schema_capsule, array_capsule);
+    }
+
+    void move(ArrowArray* array_src,
+              ArrowSchema* schema_src,
+              ArrowArray* array_dst,
+              ArrowSchema* schema_dst) {
+        memcpy(array_dst, array_src, sizeof(struct ArrowArray));
+        array_src->release = nullptr;
+
+        memcpy(schema_dst, schema_src, sizeof(struct ArrowSchema));
+        schema_src->release = nullptr;
+    }
+
+    /// \brief Call data's release callback if valid
+    void reset() {
+        if (array_.release != nullptr) {
+            array_.release(&array_);
+        }
+
+        if (schema_.release != nullptr) {
+            schema_.release(&schema_);
+        }
+    }
+
+    /// \brief Call data's release callback if valid and move ownership of the data
+    /// pointed to by data
+    void reset(ArrowArray* array_src, ArrowSchema* schema_src) {
+        reset();
+        move(array_src, schema_src, &array_, &schema_);
+    }
+
+    ~ArrowArrayHolder() {
+        reset();
+    }
+
+protected:
+    ArrowArray array_;
+    ArrowSchema schema_;
+};
+
+ArrowArrayHolder to_geoarrow(py::array_t<PyObjectGeography> input,
+                             py::object output_schema,
+                             //  py::object geometry_encoding,
+                             bool planar,
+                             float tessellate_tolerance,
+                             int precision) {
+    ArrowArrayHolder array = ArrowArrayHolder();
+
+    s2geog::geoarrow::Writer writer;
+    std::vector<std::unique_ptr<s2geog::Geography>> s2geog_vec;
+
+    s2geog::geoarrow::ExportOptions options;
+    options.set_precision(precision);
+    if (planar) {
+        auto tol = S1Angle::Radians(tessellate_tolerance / EARTH_RADIUS_METERS);
+        options.set_tessellate_tolerance(tol);
+        // TODO make this configurable
+        // options.set_projection(s2geog::geoarrow::mercator());
+    }
+
+    if (!output_schema.is(py::none())) {
+        py::capsule schema_capsule = output_schema.attr("__arrow_c_schema__")();
+        ArrowSchema* schema = static_cast<ArrowSchema*>(schema_capsule);
+        writer.Init(schema, options);
+        array.set_schema(schema);
+        // } else if (geometry_encoding.equal(py::str("WKT"))) {
+        //     writer.Init(s2geog::geoarrow::Writer::OutputType::kWKT, options);
+        // } else if (geometry_encoding.equal(py::str("WKB"))) {
+        //     writer.Init(s2geog::geoarrow::Writer::OutputType::kWKB, options);
```

Comment:
Remove? Or comment if that is for a follow-up PR?

Reply:
Yes, see my answer above, I would like to enable this, but we still lack the easy way to create the corresponding ArrowSchema if the user did not provide this. 

Will add a comment